packages/devtools_app/lib/src/flutter/table.dart

```diff
@@ -6,6 +6,68 @@ import '../table_data.dart';
 import '../trees.dart';
 import 'collapsible_mixin.dart';
 
+class FlatTable<T> extends StatefulWidget {
+  const FlatTable({
+    Key key,
+    @required this.columns,
+    @required this.data,
+    @required this.keyFactory,
+  })  : assert(columns != null),
+        assert(keyFactory != null),
+        assert(data != null),
+        super(key: key);
+
+  /// The columns to show in this table.
+  final List<ColumnData<T>> columns;
+
+  /// The tree structure of rows to show in this table.
+  final List<T> data;
+
+  /// Factory that creates keys for each row in this table.
+  ///
+  /// It uses [PageStorageKey] to restore the table's state when re-opening
+  /// its page.
+  final String Function(T data) keyFactory;
+
+  @override
+  _FlatTableState<T> createState() => _FlatTableState<T>();
+}
+
+class _FlatTableState<T> extends NestingTableState<T, FlatTable<T>> {
```

Comment:
Why does a flat table need to extend a nesting table? I would expect a "flat table" to be a table where each data elements is mapped to a single row that does not have "nesting".

Reply:
I think this is the really interesting question I want to discuss in this code review.

We have a couple of different components in play:

1. The table layout
2. The row layout
3. The nesting layout of a single cell in the tree table rows
4. Logic & animations for expanding/collapsing rows
5. Logic for turning a tree into a flat list

Right now, 1 & 5 are tightly coupled, and 2, 3 & 4 are tightly coupled.  We will already need to move 4 to be coupled with 1 to animate more efficiently when expanding/collapsing large numbers of list items (#1227).

I'm going to bang on this and find something simpler.

Reply:
To answer your specific question about why it is this way right now, the coupling between tree flattening and table layout is the answer.

Reply:
Perhaps use mixins to share common functionality but otherwise just keep the flat table completely separate from the nested table. There's already a mixin for the expand collapse animations that is used by both the nested table and the inspector tree